### DISCUSSION TITLE
How to merge two batched edge list together?

### DISCUSSION THREAD

**Discussion by CageChen** [pyg-team/pytorch_geometric#5567] [Q&A]
I have a precomputed batched edge list, and I want to learn a batched adaptive edge list in the model, how can I merge them together to input to graph conv so that the adaptive edge list can still backprop.

**Comment by EdisonLeeeee** [comment]
What's the form of your edge list? A list of `edge_index` or a `torch_geometric.data.Batch`? Can you clarify?

**Reply by CageChen** [reply]
My precomputed batched edge list is torch_geometric.data.Batch and I want to learn the nn.Parameter((batch_size, num_nodes, num_nodes) in the model. Thanks.

**Reply by EdisonLeeeee** [reply]
If I understood it correctly, you meant to convert a `torch_geometric.data.Batch` into a dense adjacency matrix with shape `(batch_size, num_nodes, num_nodes) `. You can simply use [to_dense_adj](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.to_dense_adj).

**Reply by CageChen** [reply]
学长你好，因为我感觉英文没有描述清楚，就直接用中文了哈。
我现在模型的输入是torch_geometric.data.batch，里面的edge list是我提前计算的，但是我想在模型里通过一个子网络学习得到一个(batch_size, num_nodes,num_nodes)的邻接矩阵A，并且将这个新学习的edge_list 和提前计算的edge_list 重新生成一个新的edge_list，然后送入图卷积，我该怎么做才能保证这个新学习的A得到梯度呢？

**Reply by EdisonLeeeee** [reply]
你的邻接矩阵A的值需要和你的edge_list的边权重对应，需要增加一个`edge_weight`参数指示边权重并用来计算梯度。由于A邻接矩阵包括一些不存在的边，因此你还需要扩展你的edge_list把所有边包括进来。综上，我不建议使用edge_list这种稀疏的模式进行求梯度操作，我觉得应该直接利用邻接矩阵计算会更方便一点。

**Reply by CageChen** [reply]
但是我看torch_geometric里的图卷积都是采用edge_list 的方式进行输入的，请问我是需要自己实现直接利用邻接矩阵来进行图卷积的模型嘛？

**Reply by EdisonLeeeee** [reply]
因为邻接矩阵计算效率太低，目前仅有实现一些采用邻接矩阵作为输入的layer https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#dense-convolutional-layers
如果没有你使用的，就需要你自己实现了。
